src/ServantElm.hs

```diff
@@ -0,0 +1,215 @@
+    lparen,
+    parens,
+    pretty,
+    punctuate,
+    rbrace,
+    rbracket,
+    rparen,
+    space,
+    text,
+    textStrict,
+    vsep,
+    (<$>),
+    (<+>),
+  )
+import Prelude hiding ((<$>))
+
+data LangElm
+
+instance Elm a => HasForeignType LangElm ElmDefinition a where
+  typeFor _ _ proxyA = toElmDefinition proxyA
+
+elmForAPI ::
+  ( HasForeign LangElm ElmDefinition api,
+    GenerateList
+      ElmDefinition
+      (Foreign ElmDefinition api)
+  ) =>
+  Proxy api ->
+  [Doc]
+elmForAPI api =
+  map endpointInfoToElmQuery $
+    listFromAPI (Proxy :: Proxy LangElm) (Proxy :: Proxy ElmDefinition) api
+
+elmRecord :: [Doc] -> Doc
+elmRecord = encloseSep (lbrace <> space) (line <> rbrace) (comma <> space)
+
+indent4Spaces :: Doc -> Doc
+indent4Spaces = indent 4
+
+endpointInfoToElmQuery :: Req ElmDefinition -> Doc
+endpointInfoToElmQuery requestInfo =
+  funcDef
+  where
+    funcDef =
+      vsep
+        [ fnName <+> ":" <+> typeSignature,
+          fnName <+> args <+> equals,
+          indent4Spaces elmRequest
+        ]
+
+    fnName = textStrict . camelCase $ requestInfo ^. reqFuncName
+
+    typeSignature =
+      mkTypeSignature requestInfo
+
+    args =
+      mkArgs requestInfo
+
+    elmRequest =
+      mkRequest requestInfo
+
+elmList :: [Doc] -> Doc
+elmList [] = lbracket <> rbracket
+elmList ds = lbracket <+> hsep (punctuate (line <> comma) ds) <$> rbracket
+
+stext :: Text -> Doc
+stext = text . fromStrict
+
+mkUrl :: [Segment ElmDefinition] -> Doc
+mkUrl segments =
+  urlBuilder
+    <$> (indent4Spaces . elmList)
+      (map segmentToDoc segments)
+  where
+    urlBuilder :: Doc
+    urlBuilder = "Url.Builder.crossOrigin urlBase" :: Doc
+
+    segmentToDoc :: Segment ElmDefinition -> Doc
+    segmentToDoc s =
+      case unSegment s of
+        Static path ->
+          dquotes (stext (unPathSegment path))
+        Cap arg ->
+          error
+            "TODO implement-2" -- for captures, not needed now
+
+elmTypeRefToDoc :: TypeRef -> Doc
+elmTypeRefToDoc = \case
+  RefPrim elmPrim -> elmPrimToDoc elmPrim
+  RefCustom (TypeName typeName) -> pretty typeName
+
+elmTypeParenDoc :: TypeRef -> Doc -- does it make sense to be a fn
+elmTypeParenDoc = parens . elmTypeRefToDoc
+
+elmPrimToDoc :: ElmPrim -> Doc
+elmPrimToDoc = \case
+  ElmUnit -> "()"
+  ElmNever -> "Never"
+  ElmBool -> "Bool"
+  ElmChar -> "Char"
+  ElmInt -> "Int"
+  ElmFloat -> "Float"
+  ElmString -> "String"
+  ElmTime -> "Posix"
+  ElmMaybe t -> "Maybe" <+> elmTypeParenDoc t
+  ElmResult l r -> "Result" <+> elmTypeParenDoc l <+> elmTypeParenDoc r
+  ElmPair a b -> lparen <> elmTypeRefToDoc a <> comma <+> elmTypeRefToDoc b <> rparen
+  ElmTriple a b c -> lparen <> elmTypeRefToDoc a <> comma <+> elmTypeRefToDoc b <> comma <+> elmTypeRefToDoc c <> rparen
+  ElmList l -> "List" <+> elmTypeParenDoc l
+
+mkTypeSignature :: Req ElmDefinition -> Doc
+mkTypeSignature request =
+  (hsep . punctuate " ->") ("String" : catMaybes [toMsgType, returnType])
```

Comment:
You should usually not be manually inserting spaces in strings when using a pretty printer, like in `" ->"` - shouldn't `hsep` already be taking care to put the spaces anyway?